macros/edr/data_monitoring/monitors_query/table_monitoring_query.sql

```diff
@@ -150,4 +91,182 @@
+        select edr_bucket_start,
+               edr_bucket_end,
+               start_bucket_in_data,
+               case when start_bucket_in_data is null then
+                   0
+               else {{ elementary.cast_as_float(elementary.row_count()) }} end as row_count_value
+        from buckets left join time_filtered_monitored_table on (edr_bucket_start = start_bucket_in_data)
+        group by 1,2,3
+    )
+
+    select edr_bucket_start,
+           edr_bucket_end,
+           {{ elementary.const_as_string('row_count') }} as metric_name,
+           {{ elementary.null_string() }} as source_value,
+           row_count_value as metric_value
+    from row_count_values
+{% else %}
+    select
+        {{ elementary.const_as_string('row_count') }} as metric_name,
+        {{ elementary.row_count() }} as metric_value
+    from monitored_table
+    group by 1
+{% endif %}
+{% endmacro %}
+
+{% macro freshness_metric_query(metric_args, timestamp_column=none) %}
+{% if timestamp_column %}
+    {%- set freshness_column = metric_args.freshness_column %}
+    {%- if freshness_column is undefined or freshness_column is none %}
+        {%- set freshness_column = timestamp_column %}
+    {%- endif %}
+
+    -- get ordered consecutive update timestamps in the source data
+    with unique_timestamps as (
+        select distinct {{ elementary.cast_as_timestamp(freshness_column) }} as timestamp_val
+        from monitored_table
+        order by 1
+    ),
+
+    -- compute freshness for every update as the time difference from the previous update
+    consecutive_updates_freshness as (
+        select
+            timestamp_val as update_timestamp,
+            {{ elementary.timediff('second', 'lag(timestamp_val) over (order by timestamp_val)', 'timestamp_val') }} as freshness
+        from unique_timestamps
+        where timestamp_val >= (select min(edr_bucket_start) from buckets)
+    ),
+
+    -- divide the freshness metrics above to buckets
+    bucketed_consecutive_updates_freshness as (
+        select
+            edr_bucket_start, edr_bucket_end, update_timestamp, freshness
+        from buckets, consecutive_updates_freshness
+        where update_timestamp >= edr_bucket_start AND update_timestamp < edr_bucket_end
+    ),
+
+    -- we also want to record the freshness at the end of each bucket as an additional point (as it might be larger,
+    -- or even the only point if no data was updated at all in a particular bucket)
+    bucket_end_freshness as (
+        select
+            edr_bucket_start,
+            edr_bucket_end,
+            max(timestamp_val) as update_timestamp,
+            {{ elementary.timediff('second', elementary.cast_as_timestamp('max(timestamp_val)'), "least(edr_bucket_end, {})".format(elementary.current_timestamp_column())) }} as freshness
+        from buckets, unique_timestamps
+        where timestamp_val < edr_bucket_end
+        group by 1,2
+    ),
+
+    -- create a single table with all the freshness values
```

Comment:
Should this comment be under the `consecutive_updates_freshness` CTE? just to make sure I understand it correctly.

In any case - not 100% sure about it because technically the lag can be much larger then the entire duration of the bucket (e.g. if the bucket is on the 16th and the last update was on the 14th, the freshness metric will be 2.5 days).

But maybe it's not consistent with what I did in `event_freshness` - maybe we can discuss quickly in the office